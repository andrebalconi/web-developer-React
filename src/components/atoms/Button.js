import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

export const ButtonColors = {
  default: "default",
  primary: "primary",
  danger: "danger",
};

export const ButtonsVariants = {
  default: "default",
  outlined: "outlined",
  link: "link",
};

const getMainColor = ({ theme, color }) => {
  switch (color) {
    case ButtonColors.primary:
      return theme.colors.primary.main;
    case ButtonColors.danger:
      return theme.colors.danger.main;
    default:
      return "#e0e0e0";
  }
};

const getDarkColor = ({ theme, color }) => {
  switch (color) {
    case ButtonColors.primary:
      return theme.colors.primary.dark;
    case ButtonColors.danger:
      return theme.colors.danger.dark;
    default:
      return "#5a6268";
  }
};

const getColorText = ({ theme, color }) => {
  switch (color) {
    case ButtonColors.primary:
      return theme.colors.primary.text;
    case ButtonColors.danger:
      return theme.colors.danger.text;
    default:
      return "#212121";
  }
};

const getOutlinedText = (props) => {
  if (props.color === ButtonColors.default) {
    return "#212121";
  }

  return getMainColor(props);
};

const getLinkText = (props) => {
  if (props.color === ButtonColors.default) {
    return "#757575";
  }

  return getMainColor(props);
};

const Button = styled.button`
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: ${getMainColor};
  border: 2px solid ${getMainColor};
  color: ${getColorText};
  display: inline-block;
  text-decoration: none;

  &:disabled {
    opacity: 0.6;
    cursor: not-allowed;
  }

  &:hover:not(:disabled) {
    background-color: ${getDarkColor};
    border-color: ${getDarkColor};
  }
`;

const ButtonOutlined = styled(Button)`
  background-color: transparent;
  color: ${getOutlinedText};

  &:hover:not(:disabled) {
    background-color: transparent;
    color: ${getDarkColor};
  }
`;

const ButtonLink = styled(Button)`
  background-color: transparent;
  border-color: transparent;
  color: ${getLinkText};
  padding-left: 0;
  padding-right: 0;

  &:hover:not(:disabled) {
    background-color: transparent;
    border-color: transparent;
    color: ${getDarkColor};
  }
`;

const ButtonWrapper = (props) => {
  switch (props.variant) {
    case ButtonsVariants.outlined:
      return <ButtonOutlined {...props} />;
    case ButtonsVariants.link:
      return <ButtonLink {...props} />;
    default:
      return <Button {...props} />;
  }
};

ButtonWrapper.defaultProps = {
  type: "button",
  children: undefined,
  color: "default",
  variant: "default",
};

ButtonWrapper.propTypes = {
  type: PropTypes.string,
  children: PropTypes.node,
  color: PropTypes.oneOf(Object.values(ButtonColors)),
  variant: PropTypes.oneOf(Object.values(ButtonsVariants)),
};

export default ButtonWrapper;
