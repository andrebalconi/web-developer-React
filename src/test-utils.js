import React from "react";
import ThemeProvider from "styles/ThemeProvider";
import { render } from "@testing-library/react";
/*import { MemoryRouter as Router } from "react-router-dom";*/

const TestProvider = ({ children }) => (
  
    <ThemeProvider>{children}</ThemeProvider>

);

export * from "@testing-library/react";

const customRender = (ui, options) =>
  render(ui, { wrapper: TestProvider, ...options });

export { customRender as render };
