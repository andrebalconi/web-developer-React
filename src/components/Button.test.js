import React from "react";
import { render, fireEvent } from "test-utils";

import Button, { ButtonColors, ButtonsVariants } from "./Button";

test("renders a text", () => {
  const { getByText } = render(<Button>Click here</Button>);

  expect(getByText("Click here")).toBeInTheDocument();
});

test("trigger event on click", () => {
  const handleClick = jest.fn();

  const { getByRole } = render(<Button onClick={handleClick} />);

  fireEvent.click(getByRole("button"));

  expect(handleClick).toBeCalled();
});

test.each(Object.values(ButtonColors).map((item) => [item]))(
  "renders with color %s",
  (color) => {
    const { asFragment } = render(<Button color={color} />);

    expect(asFragment()).toMatchSnapshot();
  }
);

test.each(Object.values(ButtonsVariants).map((item) => [item]))(
  "renders with variant %s",
  (variant) => {
    const { asFragment } = render(<Button variant={variant} />);

    expect(asFragment()).toMatchSnapshot();
  }
);

test("renders with color primary and variant outlined", () => {
  const { asFragment } = render(<Button variant="outlined" color="primary" />);

  expect(asFragment()).toMatchSnapshot();
});

test("renders with color primary and variant link", () => {
  const { asFragment } = render(<Button variant="link" color="primary" />);

  expect(asFragment()).toMatchSnapshot();
});
