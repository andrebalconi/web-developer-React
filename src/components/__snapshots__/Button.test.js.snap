// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with color danger 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #ff3d00;
  border: 2px solid #ff3d00;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #b22a00;
  border-color: #b22a00;
}

<button
    class="c0"
    color="danger"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with color default 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

<button
    class="c0"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with color primary 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #ffc107;
  border: 2px solid #ffc107;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #c79100;
  border-color: #c79100;
}

<button
    class="c0"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with color primary and variant link 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #ffc107;
  border: 2px solid #ffc107;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #c79100;
  border-color: #c79100;
}

.c1 {
  background-color: transparent;
  border-color: transparent;
  color: #ffc107;
  padding-left: 0;
  padding-right: 0;
}

.c1:hover:not(:disabled) {
  background-color: transparent;
  border-color: transparent;
  color: #c79100;
}

<button
    class="c0 c1"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with color primary and variant outlined 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #ffc107;
  border: 2px solid #ffc107;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #c79100;
  border-color: #c79100;
}

.c1 {
  background-color: transparent;
  color: #ffc107;
}

.c1:hover:not(:disabled) {
  background-color: transparent;
  color: #c79100;
}

<button
    class="c0 c1"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with variant default 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

<button
    class="c0"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with variant link 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

.c1 {
  background-color: transparent;
  border-color: transparent;
  color: #757575;
  padding-left: 0;
  padding-right: 0;
}

.c1:hover:not(:disabled) {
  background-color: transparent;
  border-color: transparent;
  color: #5a6268;
}

<button
    class="c0 c1"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with variant outlined 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

.c1 {
  background-color: transparent;
  color: #212121;
}

.c1:hover:not(:disabled) {
  background-color: transparent;
  color: #5a6268;
}

<button
    class="c0 c1"
    color="default"
    type="button"
  />
</DocumentFragment>
`;
